_posts/2019-04-23-js-recon-is-no-more.md

```diff
@@ -0,0 +1,147 @@
+with related sections. I expect you run code samples in the developer console of
+your browser.
+
+### Glossary
+
+* **Empty / Available port**: A port where no service is running.
+* **Non-empty / Occupied port**: A port where some service is running.
+* **XHR**: A short form of [XML Http Request][mdn_xhr].
+* **Socket**: A TCP/IP raw socket.
+
+According to that blog post, If I write a Javascript code to open an XHR to
+`http://localhost:8084` and host that code at example.com then when you visit
+the `example.com` the browser will open that XHR to port `8084` of your
+workstation, because the `localhost` for your browser is your workstation. This
+gap invites many vulnerabilities for users. One of them is the possibility to
+fingerprint open TCP ports at the client workstation.
+
+That post further claims that the browser takes recognizably more time to open
+an XHR targeting an occupied port. Comparatively, time took to open an XHR aimed
+at an empty port was short.
+
+```javascript
+//Sample code to measure the time browser took to open the socket.
+
+var requestPort = function(port) {
+  var startTime = null;
+  var xhr = new XMLHttpRequest();
+  xhr.onreadystatechange = function() {
+    if (xhr.readyState === 1) {
+      var timeTook = Date.now() - startTime;
+      console.log("Browser took : " + timeTook + " microseconds to open.");
+    }
+  };
+  startTime = Date.now();
+  url = "http://localhost:" + port;
+  xhr.open("GET", url, true);
+};
+```
+
+You can paste this code at the developer console of your browser. Calling this
+function by writing `requestPort(8084)` at the console will open an XHR for port
+`8084`. The function will print how long the browser took to open a socket on
+that port. I invite you to call this function with a combination of empty and
+non-empty ports to observe various response timings.
+
+I tried opening a bunch of requests using `requestPort()` function at suspected
+ports. For me the method was giving unidentifiable patterns in the time the
+browser took to open a socket. Below is a histogram of comparing times taken to
+open a socket on empty port(8084) and non-empty port(27017).
+
+![Graph showing response time to open socket.]({{site.url}}/assets/images/js_recon_is_no_more/graph_response_time_open_socket.png)
+
+
+Above is a graph of 6000 requests done to measure the response times using the
+function I shared earlier. More than 5000 requests have ended in nearly no time.
+There were less than 1000 requests which ended in 1 microseconds. From the above
+results, We can conclude that mentioned method is not giving different results
+for occupied and empty port. We can conclude that mentioned method by AndLabs is
+failing to distinguish occupied ports from nonoccupied ports.
+
+I tried hard to find possible causes for the failure of this attack. I didn't
+find any conclusive evidence. Perhaps our hardware or browser code has improved
```

Comment:
See my comment about networking. Generally speaking networking requests are slow, yet you claim to find meaningful differences between a single microsecond and "no time at all", so I'd start there...